### PR TITLE
[Parse] Enable self rebinding (to self)

### PR DESCRIPTION
**Pull request by dduan**
As discussed in this [thread](https://forums.swift.org/t/the-future-of-weak-self-rebinding/10846), this patch allows "self" to be an identifier bound to `self` (see added tests for examples), and is useful in closures where `self` is weakly captured to avoid retain cycles. The current behavior of allowing 

```swift
guard let `self` = self else { ... }
```

remains for source compatibility.

Instead of parsing identifiers and fallthrough to parsing self, treat `kw_self` the same as identifiers.

### CODE REVIEW

**Comment by dduan** [comment]
@swift-ci please test

**Comment by rintaro** [comment]
Could you add a test cases for implicit method/property references?

```swift
class MyCls {
  func something() {}

  func test() {
    let self = OtherCls() // Even if `self` are shadowed,
    something() // this should still refer `MyCls.something`.
  }
}
```
I know this currently works. But it prevents accidental breakage.

**Comment by dduan** [comment]
@swift-ci please test

**Comment by dduan** [comment]
@rintaro added

**Comment by dduan** [comment]
@swift-ci please test

**Comment by dduan** [comment]
@rintaro does everything else look mergeable to you? cc @DougGregor (owner)

**Comment by rintaro** [comment]
No more objection from me. Defer to @DougGregor (or other core team member) on the decision.

File: test/Parse/self_rebinding.swift
Comment (DougGregor) [inline test/Parse/self_rebinding.swift]: It'd be nice to have some part of the warning text here, for others that read the test later on. Is the warning about rebinding `self` to an unrelated type (which would be a really good warning, IMO!), or is it something else?

File: test/Parse/self_rebinding.swift
Comment (DougGregor) [inline test/Parse/self_rebinding.swift]: You can delete this line; it doesn't add anything.

**Review by DougGregor — APPROVED** [review]

File: test/Parse/self_rebinding.swift
Comment (dduan) [reply test/Parse/self_rebinding.swift]: in this case it's just about the variable being unused. I included the text.

**Comment by dduan** [comment]
@swift-ci please smoke test

**Comment by dduan** [comment]
@swift-ci please smoke test

**Comment by dduan** [comment]
@swift-ci please test

**Comment by acecilia** [comment]
Is this PR linked to an evolution proposal?

There is an old evolution proposal [here](https://github.com/apple/swift-evolution/blob/master/proposals/0079-upgrade-self-from-weak-to-strong.md) already merged, but in **Deferred** status. Seen the general acceptance shown in the [discussion thread](https://forums.swift.org/t/the-future-of-weak-self-rebinding/10846) + the merge of this PR, would it make sense to consider updating the evolution proposal and taking it back to life?

**Comment by lattner** [comment]
@dduan and @DougGregor can you two look at revising and updating SE-0079?  It is related to this patch and is causing confusion for people who think that this behavior is not supposed to be supported.

**Comment by dduan** [comment]
Will a simple update of status suffice? https://github.com/apple/swift-evolution/pull/900

**Pull request merged**
